_posts/2025-02-01-next-scala-lts.md

```diff
@@ -0,0 +1,82 @@
+---
+category: announcement
+permalink: /news/next-scala-lts.html
+title: "Next Scala 3 LTS version"
+by: Tomasz Godzik
+---
+
+**TLDR;**
+
+sun.misc.Unsafe, re-used in lazy val implementation, will be removed in a future
+JDK one of the Scala 3 Next minors in 2025 Q4 and the next LTS will drop JDK 8
+support
+
+## Intro
+
+For the compiler’s second LTS version, the Scala 3 compiler team will be
+dropping support for JDK 8 and is considering which later JDK to use.
+
+## Reasons for the change
+
+One of the main reasons is that recent [JEP 471](https://openjdk.org/jeps/471)
+stated that the memory-access methods in sun.misc.Unsafe are scheduled for
+removal in a future release. Currently, Scala 3 uses sun.misc.Unsafe in its
+implementation of lazy values. This was needed due to compatibility with JDK 8.
+In order to support later versions of JDK we will need to drop usage of Unsafe,
+which is being investigated under
+[this issue](https://github.com/scala/scala3/issues/9013). It also started to
+dawn on the compiler team that we might consider stopping to support JDK 8
+altogether in a future release of Scala 3.
+
+## Advantages
+
+Switching to a newer version of the JDK would allow both the compiler and the
+Scala ecosystem to start using new features brought in by JDK 9+. This might
+include a number of JEPs, which will be investigated by the team the coming
+year. To check the current status take a look at the issues marked with the
+[JEP label](https://github.com/scala/scala3/issues?q=is%3Aissue+is%3Aopen+label%3Acompat%3Ajava%3Ajep).
```

Comment:
in addition to providing the link, I would suggest briefly highlighting 2 or 3 of the most important ones. Just giving the link has a sort of "you figure it out, we can't be bothered to summarize it for you" vibe

Reply:
I added some, but I am not sure how well I understand them

Reply:
heh, well me either, maybe @sjrd would like to take a look?

Reply:
I would remove JEP 334. I don't think this is useful to us.

We could mention JEP 409 (sealed classes). Similarly to nest-based access control, we could produce bytecode with restrictions that better match Scala's language-level restrictions. Not a big deal, but there is some use case for it.

Looking further into the future, the addition of [`SwitchBootstraps`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/runtime/SwitchBootstraps.html) in JDK 21 *might* allow us to encode long pattern matches more efficiently for the JVM.

Reply:
Thanks, I will switch 334 to 409. Should we open an issue for SwitchBootstraps? Was that part of a JEP or added independently?

Reply:
Ok, I found that it's 441 (at least it looks like part of it) and added it to the list

Reply:
I  told @liach about this, and the jvm team here, they suggested to Java 21 as a new foundation,  the Java 21 will be the new Java 8 in next N years, and Java 25 the next LTS have a fix for Virtual Threads, which will accelerate the adoption too.

At work,  mostly we are just migrating our system from Java 8 to Java 21, and the jvm team are putting more energy in Java 21 and Java 11.

Move to Java 21 help Scala on pattern matching with records too, and the language can leverage VT, typeswitch , zgc things which are added in Java 21.

Reply:
Makes sense! We can discuss it further once the blogpost is out and we have a contributors thread. This blogpost is just to announce that we are moving towards a newer jdk

Reply:
https://javaalmanac.io for api comparing.

Reply:
For bumping, I think the most significant advantages in the JVM may be:
1. JEP 181 Nestmates, to eliminate bridge methods 
2. JEP 309 Condy, more simple representation of bootstrap method/indy args
3. JEP 371 Hidden classes, useful to enclose methods generated to implement indy call sites (such as dynamic implementation for pattern matching)

Reply:
@liach these are really examples, there are no actual decisions yet in this document. It's worth to make sure those issues exist in Scala 3 repo